### DISCUSSION TITLE
update-citations errors "on push to main"

### DISCUSSION THREAD

**Discussion by imnanxu** [greenelab/lab-website-template#284] [Q&A] [answered]
### Checks

- [X] I have searched **[the docs](https://greene-lab.gitbook.io/lab-website-template-docs)**, [existing issues](https://github.com/greenelab/lab-website-template/issues), and [existing discussions](https://github.com/greenelab/lab-website-template/discussions) for answers first.

### Link to your website repo

https://github.com/inspirelab-site/inspirelab.site

### Version of Lab Website Template you are using

1.2.2

### Description

I updated some bio information without updating the citations, and everything ran fine locally. However, after I pushed to main on GitHub, it keeps having the update-citation errors: https://github.com/inspirelab-site/inspirelab.site/actions/runs/11307108274
<img width="1158" alt="Screenshot 2024-10-12 at 12 39 07 PM" src="https://github.com/user-attachments/assets/629c2212-3fd0-497d-bcbc-66214c63f2f8">

This happens quite often lately. The same error could appear on a different citation(s) if I hit re-run jobs. Sometimes, it resolved by hitting re-run the jobs for multiple times, but this time, it doesn't. I don't know how to fix it as all citations are running fine locally.

**Comment by vincerubinetti** [comment, marked as answer]
This is almost certainly an upstream either with Manubot, or one of its dependencies. But I can try to help.

It could be that the `eutils` API (as you see in your error logs) has just been spotty lately. Or I wonder if somehow you are being rate-limited or otherwise blocked by the API for making too many requests.

Same as you, I can't replicate the issue locally. On Monday I'll try running some tests from an actual GitHub Actions environment to see if I can reproduce.

> all citations are running fine locally.

For now, I'd recommend running the template locally and just committing the citation results generated there.

---

Also, I think you're already aware of this, but looking at the failures in your workflow runs, many are actually due to Ruby package errors. I seem to recall in your site you've modified lots of things from the template, including `Gemfile.lock`, which could cause issues. It will be difficult to help you with these issues due to the modifications.

**Reply by imnanxu** [reply]
Thanks for the quick follow-up! Just to clarify, I did make some modifications, but only in the _style files like portrait.scss and in _layouts like member.html. I haven’t touched anything related to the Gemfile.lock or other dependencies, so I’m not sure if that’s contributing to the issue.

I did follow the Security warnings to update some dependencies like "bundler in for nokogiri" and "bundler in for rexml" just tonight, but that hasn't resolved the issues... OK. I’ll still keep an eye on the eutils API and will try running things locally as you suggested. Let me know if there’s anything else I should try in the meantime!

**Reply by imnanxu** [reply]
I've noticed that the entries causing errors consistently are the ones called by PMID. In my sources.yaml file, I changed all PMIDs to DOIs, and also replaced the files in _cite and _cite/plugins to some old versions, which seem to fix the citation issues. I wonder if there's a way to display both PMID and DOI as well as PMC at the end of each publication.

Now, the issue stuck on "build-site"-->"Install Ruby Packages", which I don't know how to fix. I've also replaced the Gemfile.lock to the original version in case it was unintentionally changed by the Security warning updates... Now, you can see all security warnings (Dependabot alerts) came back...

 Sorry I have pushed the changes to the main for test.
<img width="1289" alt="Screenshot 2024-10-13 at 12 03 33 AM" src="https://github.com/user-attachments/assets/e68becb8-0f84-44c4-87c9-9a051d3f622a">
>

**Reply by vincerubinetti** [reply]
> I've noticed that the entries causing errors consistently are the ones called by PMID

This is what makes me certain it is an upstream issue. The template itself is not aware of doi vs pmid vs whatever, Manubot handles all of that. Will look into this today.

> also replaced the files in _cite and _cite/plugins to some old versions

I'd recommend keeping the latest versions of the files. Having older versions may create other errors that are hard for me to troubleshoot.

> I wonder if there's a way to display both PMID and DOI as well as PMC at the end of each publication.

You could put the PMID in the `description` field and it will show up automatically, though it'll be in its own paragraph. You could also use the [`buttons` field
](https://greene-lab.gitbook.io/lab-website-template-docs/basics/citations#rich-details) too, though that will also take up extra space. To put it on the same line as the main ID, you could edit `_includes/citation.html` to display an `alt-id` field, then manually assign `alt-id: PMID:123456` in your sources.

> Now, the issue stuck on "build-site"-->"Install Ruby Packages", 

This is a separate issue, please follow this issue here:

https://github.com/greenelab/lab-website-template/issues/285

**Reply by vincerubinetti** [reply]
Actually, I am now able to see the issue locally sometimes, after upgrading to the latest Manubot and manually running `manubot cite pubmed:28559793`, and even just running a request to eutils directly.

This upstream issue on Manubot I created here for more details:
https://github.com/manubot/manubot/issues/374

Unfortunately I'm not sure there's much we can do. It really seems to be either flaky-ness or rate-limiting in the eutils API. Given that, the solution would be to automatically retry a request when this happens.

For now, I'm going to wait for the Manubot folks to to implement retries. If they haven't been able to do that after say a few months, I'll think about implementing it in the template `cite.py` script.

In the meantime, I'd recommend using DOIs whenever possible, as anecdotally they seem to be more reliable, and just retry when a job fails. I plan to make failure retries on GitHub Actions faster soon, see this issue: https://github.com/greenelab/lab-website-template/issues/287
